lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1521 @@
+}
+
+/// From a domain info attribute, get the domain-type of a domain value at
+/// index i.
+StringAttr getDomainPortTypeName(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return nullptr;
+  auto ref = cast<FlatSymbolRefAttr>(info[i]);
+  return ref.getAttr();
+}
+
+/// From a domain info attribute, get the row of associated domains for a
+/// hardware value at index i.
+auto getPortDomainAssociation(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return info.getAsRange<IntegerAttr>();
+  return cast<ArrayAttr>(info[i]).getAsRange<IntegerAttr>();
+}
+
+/// Return true if the value is a port on the module.
+bool isPort(BlockArgument arg) {
+  return isa<FModuleOp>(arg.getOwner()->getParentOp());
+}
+
+/// Return true if the value is a port on the module.
+bool isPort(Value value) {
+  auto arg = dyn_cast<BlockArgument>(value);
+  if (!arg)
+    return false;
+  return isPort(arg);
+}
+
+/// Returns true if the value is driven by a connect op.
+bool isDriven(Value port) {
+  for (auto *user : port.getUsers())
+    if (auto connect = dyn_cast<FConnectLike>(user))
+      if (connect.getDest() == port)
+        return true;
+  return false;
+}
+
+//====--------------------------------------------------------------------------
+// Global State.
+//====--------------------------------------------------------------------------
+
+/// Each domain type declared in the circuit is assigned a type-id, based on the
+/// order of declaration. Domain associations for hardware values are
+/// represented as a list, or row, of domains. The domains in a row are ordered
+/// according to their type's id.
+struct DomainTypeID {
+  size_t index;
+};
+
+/// Information about the domains in the circuit. Able to map domains to their
+/// type ID, which in this pass is the canonical way to reference the type
+/// of a domain, as well as provide fast access to domain ops
+class DomainInfo {
+public:
+  DomainInfo(CircuitOp circuit) { processCircuit(circuit); }
+
+  ArrayRef<DomainOp> getDomains() const { return domainTable; }
+  size_t getNumDomains() const { return domainTable.size(); }
+  DomainOp getDomain(DomainTypeID id) const { return domainTable[id.index]; }
+  // DomainOp getDomain(DomainType type) const {
+  //   return getDomain(getDomainTypeID(type));
+  // }
+  DomainTypeID getDomainTypeID(StringAttr name) const {
+    return typeIDTable.at(name);
+  }
+
+  DomainTypeID getDomainTypeID(FlatSymbolRefAttr ref) const {
+    return getDomainTypeID(ref.getAttr());
+  }
+
+  DomainTypeID getDomainTypeID(ArrayAttr info, size_t i) const {
+    auto name = getDomainPortTypeName(info, i);
+    return getDomainTypeID(name);
+  }
+
+  DomainTypeID getDomainTypeID(Value value) const {
+    assert(isa<DomainType>(value.getType()));
+    if (auto arg = dyn_cast<BlockArgument>(value)) {
+      auto *block = arg.getOwner();
+      auto *owner = block->getParentOp();
+      auto module = cast<FModuleOp>(owner);
```

Comment:
We've had a few commits moving us off of use of `module` in our codebase as it's special starting with C++20.

GitHub's syntax highlighting thinks it's a keyword.